cloud-controller-manager/do/certificates_test.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2017 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/digitalocean/godo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+type kvCertService struct {
+	store         map[string]*godo.Certificate
+	reflectorMode bool
+}
+
+func (f *kvCertService) Get(ctx context.Context, certID string) (*godo.Certificate, *godo.Response, error) {
+	if f.reflectorMode {
+		return &godo.Certificate{
+			ID:   certID,
+			Type: certTypeLetsEncrypt,
+		}, newFakeOKResponse(), nil
+	}
+	lb, ok := f.store[certID]
+	if ok {
+		return lb, newFakeOKResponse(), nil
+	}
+	return nil, newFakeNotOKResponse(), newFakeNotFoundErrorResponse()
+}
+
+func (f *kvCertService) List(ctx context.Context, listOpts *godo.ListOptions) ([]godo.Certificate, *godo.Response, error) {
+	panic("not implemented")
+}
+
+func (f *kvCertService) Create(ctx context.Context, crtr *godo.CertificateRequest) (*godo.Certificate, *godo.Response, error) {
+	panic("not implemented")
+}
+
+func (f *kvCertService) Delete(ctx context.Context, certID string) (*godo.Response, error) {
+	panic("not implemented")
+}
+
+func newKVCertService(store map[string]*godo.Certificate, reflectorMode bool) kvCertService {
+	return kvCertService{
+		store:         store,
+		reflectorMode: reflectorMode,
+	}
+}
+
+func createServiceAndCert(lbID, certID, certType string) (*v1.Service, *godo.Certificate) {
```

Comment:
refactor: I implemented a [service builder](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/67c685a86ca3b54897fa4f835c0e652844466a9a/cloud-controller-manager/do/resources_test.go#L42) elsewhere which could be extended and leveraged for testing certificates.

For common service builder parameters, I also [created shortcuts at times](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/67c685a86ca3b54897fa4f835c0e652844466a9a/cloud-controller-manager/do/resources_test.go#L94-L96).